File: 卡码/代码随想录/1/09-动态规划/C++/dp/02-背包问题/src/02-完全背包/LeetCode_0279.cpp
//279. 完全平方数
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int numSquares(int n) {
		vector<int> dp(n+1,INT_MAX-1);
		dp[0]=0;
		for(int i=1;i*i<=n;i++){
			for(int j=i*i;j<=n;j++){
				dp[j]=min(dp[j],dp[j-i*i]+1);
			}
		}
		return dp[n];
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/09-动态规划/Java/dp/src/main/java/com/kurumi/dp/_02_背包问题/_02_完全背包/LeetCode_0279.java
//279. 完全平方数
package com.kurumi.dp._02_背包问题._02_完全背包;

import java.util.Arrays;

public class LeetCode_0279 {

}
class Solution {
  public int numSquares(int n) {
    int[] dp=new int[n+1];
    Arrays.fill(dp,Integer.MAX_VALUE-1);
    dp[0]=0;
    for(int i=1;i*i<=n;i++){
      for(int j=i*i;j<=n;j++){
        dp[j]=Math.min(dp[j], dp[j-i*i]+1);
      }
    }
    return dp[n];
  }
}
